ohttp-client/src/lib.rs

```diff
@@ -0,0 +1,438 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use bhttp::{Message, Mode};
+use futures_util::stream::unfold;
+use ohttp::ClientRequest;
+use rand::distributions::{Alphanumeric, DistString};
+use reqwest::{Client, Response};
+use serde::Deserialize;
+use std::{
+    fs::{self, File},
+    io::{Cursor, Read, Write},
+    ops::Deref,
+    path::PathBuf,
+    str::FromStr,
+};
+use tracing::{error, info, trace};
+use warp::hyper::body::Body;
+
+type Res<T> = Result<T, Box<dyn std::error::Error>>;
+
+#[derive(Debug, Clone)]
+/// This allows a `HexArg` to be created from a string slice (`&str`) by decoding
+/// the string as hexadecimal.
+pub struct HexArg(Vec<u8>);
+impl FromStr for HexArg {
+    type Err = hex::FromHexError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        hex::decode(s).map(HexArg)
+    }
+}
+
+/// This allows `HexArg` instances to be dereferenced to a slice of bytes (`[u8]`).
+impl Deref for HexArg {
+    type Target = [u8];
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+/// Writes the request line for an HTTP POST request to the provided buffer.
+/// The request line follows the format:
+/// `POST {target_path} HTTP/1.1\r\n`.
+fn write_post_request_line(request: &mut Vec<u8>, target_path: &str) -> Res<()> {
+    write!(request, "POST {target_path} HTTP/1.1\r\n")?;
+    Ok(())
+}
+
+/// Appends HTTP headers to the provided request buffer.
+fn append_headers(request: &mut Vec<u8>, headers: &Vec<String>) -> Res<()> {
+    for header in headers {
+        write!(request, "{header}\r\n")?;
+        info!("{header}\r\n");
+    }
+    Ok(())
+}
+
+/// Creates a multipart/form-data body for an HTTP request.
```

Comment:
I had to drop them because pyo3 toolchain was giving some errors with these comments. I can investigate this and add the comments back at some point.